### PR TITLE
[ASCII-1281] Ensure host metadata information information is sent first 

### PR DESCRIPTION
**Pull request by GustavoCaso**
### What does this PR do?

We add a new  `MetadataPriorityProvider`. This allow us to ensure the host metadata information is sent synchronously at start time. This is a workaround to an issue in the backend regarding [Metrics Tags Delay](https://datadoghq.atlassian.net/wiki/spaces/METRICSIN/pages/3082259924/Host+Tags+Delays) 

### Motivation

Workaround for incident https://app.datadoghq.com/incidents/25790

### Possible Drawbacks / Trade-offs

⚠️ **This change mitigates a race condition issue with the backend, but does not solve the problem 100% of the times. When doing the QA be aware that there might be cases in which the host tags are delayed** ⚠️ 

<details>
  <summary>Here is summary of the findings during the investigation of the incident</summary>
 The backend has multiple endpoints to consume different host information and create the necessary records in the database. The backend also has various ways to create a host entry in the DB.

- Via the `intake` endpoint. The data is stored in the PostgreSQL DB. The agents send the host tags information (agent V5 payload) and the system events.

- Via the `metadata` endpoint. This endpoint stores the information REDAPL.  The agent sends the metadata information to this endpoint (agent metadata, check metadata and host metadata). To our surprise, we found out that this endpoint also creates a host record in the PostgreSQL DB if not present.

Another essential thing to remember about the `intake` endpoint is that when a request is made and a host entry is read, what is read from the DB is stored in the **cache for 5 minutes**. This is to avoid doing too many reads in the DB.
In version 7.50, the agent first issues the host payload (agent V5 payload) to the  intake endpoint at startup. The payloads to the `metadata` endpoint were sent 1 minute after the agent startup. Here is the list of requests with their timestamp.

```
127.0.0.1 - - [14/Mar/2024 18:14:13] "GET /api/v1/validate?api_key=xxxxxxxxxx HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:14:19] "POST /intake/ HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:14:28] "POST /intake/ HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:14:28] "POST /api/v1/check_run HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:14:28] "POST /api/v2/series HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:14:43] "POST /api/v1/check_run HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:14:43] "POST /api/v2/series HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:14:58] "POST /api/v1/check_run HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:14:58] "POST /api/v2/series HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:15:13] "POST /api/v1/check_run HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:15:13] "POST /api/v2/series HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:15:13] "POST /api/v1/metadata HTTP/1.1" 200 -
```

Here, the first request to the intake creates the host record, and the second request to the intake endpoint is stored what is the DB in the cache.

The changes in 7.51 https://github.com/DataDog/datadog-agent/pull/20234 and https://github.com/DataDog/datadog-agent/pull/21146 relate to how the agent metadata and host payload are sent to the backend. The most significant change is that at startup, we spin up a bunch of goroutines, one per metadata provider (host tags, check metadata, agent metadata, and host metadata) and send the data to the backend.

```
127.0.0.1 - - [14/Mar/2024 18:18:08] "GET /api/v1/validate?api_key=xxxxxxxxxxx HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:08] "POST /api/v1/metadata HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:08] "POST /api/v1/metadata HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:14] "POST /api/v1/metadata HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:14] "POST /intake/ HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:23] "POST /intake/ HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:23] "POST /api/v1/check_run HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:23] "POST /api/v2/series HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:38] "POST /api/v1/check_run HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:38] "POST /api/v2/series HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:53] "POST /api/v1/check_run HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 18:18:53] "POST /api/v2/series HTTP/1.1" 200 -
```

The order in which the request is executed is random.
Here, the first request to the metadata creates the host record, and the first request to the intake endpoint is stored in the cache. This results in the cached result not having the tags.
That's why we saw an increase in tag delay in 7.51.0. Still, the delay is 5 minutes, when the cache expires and gets populated with the latest data.
The fix we suggested for 7.51.2 forced the intake payload to be sent first.

```
127.0.0.1 - - [14/Mar/2024 15:24:00] "GET /api/v1/validate?
127.0.0.1 - - [14/Mar/2024 15:24:07] "POST /intake/ HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 15:24:07] "POST /api/v1/metadata HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 15:24:07] "POST /api/v1/metadata HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 15:24:07] "POST /api/v1/metadata HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 15:24:07] "POST /intake/ HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 15:24:15] "POST /intake/ HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 15:24:15] "POST /api/v1/check_run HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 15:24:15] "POST /api/v2/series HTTP/1.1" 200 -
127.0.0.1 - - [14/Mar/2024 15:24:30] "POST /api/v1/check_run HTTP/1.1" 200 -
```

It mitigates the issue but does not solve the race condition in the backend. Since most of the requests are fired simultaneously, it is a matter of which backend service finishes processing the request first.
Since it is a race condition issue, the changes made in 7.51.0 do not affect all customers. Customers who tend to redeploy or restart the agent more often have a higher chance of hitting this race condition.
</details>

### Describe how to test/QA your changes

Ensure agent metrics have tags are present from the beginning. 
You can add system tags into the datadog configuration in the `tags` . We can add custom tags in the `datadog.yaml` file
Ex. `tags: ["foo:bar", "hello:world"]` 

To test that metrics do not have `N/A` tags. Run the agent with some custom tags. 
1. Ensure your agent shows in the UI under the infrastructure list 
Screenshot 2024-03-11 at 16 25 09
2. Go to the metrics explored and validate that when querying for our custom tags we do not get `nil` values
image

💡 In the screenshot below we can compare two agent running. One with the fix the other running from main.
image
We can see the one in main we see metrics with tags `N/A`

### CODE REVIEW

File: comp/metadata/runner/runnerimpl/runner.go [hunk 1]
```diff
@@ -72,11 +92,30 @@ func NewEmptyProvider() Provider {
...
+		if p, isSet := optionaP.Get(); isSet {
+			providers = append(providers, p)
+		}
+	}
+
+	for _, optionaP := range nonNilPriorityProviders {
+		if p, isSet := optionaP.Get(); isSet {
+			priorityProviders = append(priorityProviders, p)
+		}
+	}
```
Comment (GustavoCaso) [inline comp/metadata/runner/runnerimpl/runner.go:111]: I moved this functionality here mostly to leave the `start` function clean and to report correct number of providers.

The log line `r.log.Debugf("Starting metadata runner with %d providers", len(r.providers)+len(r.priorityProviders))` was reporting wrong numbers as it was also counting optional providers

File: comp/metadata/runner/runnerimpl/runner.go [hunk 2]
```diff
@@ -137,12 +176,21 @@ func (r *runnerImpl) handleProvider(p MetadataProvider) {
 // start is called by FX when the application starts. Lifecycle hooks are blocking and called sequencially. We should
 // not block here.
 func (r *runnerImpl) start() error {
-	r.log.Debugf("Starting metadata runner with %d providers", len(r.providers))
-	for _, optionaP := range r.providers {
-		if p, isSet := optionaP.Get(); isSet {
-			go r.handleProvider(p)
+	r.log.Debugf("Starting metadata runner with %d providers", len(r.providers)+len(r.priorityProviders))
+
+	go func() {
```
Comment (GustavoCaso) [inline comp/metadata/runner/runnerimpl/runner.go:181]: By wrapping the functionality of initializing the provides we ensure not blocking the application start.

File: comp/metadata/runner/runnerimpl/runner.go [hunk 3]
```diff
@@ -27,13 +27,18 @@ func Module() fxutil.Module {
 // MetadataProvider is the provider for metadata
 type MetadataProvider func(context.Context) time.Duration
 
+// PriorityMetadataProvider is the provider for metadata that needs to be execute at start time of the agent.
+// Right now the agent needs the host metadata provider to execute first to ensure metrcis tags are being reported correctly.
```
Comment (hush-hush) [inline comp/metadata/runner/runnerimpl/runner.go:31]:
```suggestion
// Right now the agent needs the host metadata provider to execute first to ensure host tags are being reported correctly.
```

File: comp/metadata/runner/runnerimpl/runner.go [hunk 4]
```diff
@@ -72,11 +92,30 @@ func NewEmptyProvider() Provider {
 
 // createRunner instantiates a runner object
 func createRunner(deps dependencies) *runnerImpl {
+	providers := []MetadataProvider{}
+	priorityProviders := []PriorityMetadataProvider{}
+
+	nonNilProviders := fxutil.GetAndFilterGroup(deps.Providers)
+	nonNilPriorityProviders := fxutil.GetAndFilterGroup(deps.PriorityProviders)
```
Comment (hush-hush) [inline comp/metadata/runner/runnerimpl/runner.go:99]: do we need the `nonNil` part in the name ?

File: comp/metadata/runner/runnerimpl/runner.go (hunk 2 above)
Comment (hush-hush) [reply comp/metadata/runner/runnerimpl/runner.go:181]: We should add a comment explaining this.

File: comp/metadata/runner/runnerimpl/runner.go [hunk 5]
```diff
@@ -137,12 +176,21 @@ func (r *runnerImpl) handleProvider(p MetadataProvider) {
 // start is called by FX when the application starts. Lifecycle hooks are blocking and called sequencially. We should
 // not block here.
 func (r *runnerImpl) start() error {
-	r.log.Debugf("Starting metadata runner with %d providers", len(r.providers))
-	for _, optionaP := range r.providers {
-		if p, isSet := optionaP.Get(); isSet {
-			go r.handleProvider(p)
+	r.log.Debugf("Starting metadata runner with %d providers", len(r.providers)+len(r.priorityProviders))
```
Comment (hush-hush) [inline comp/metadata/runner/runnerimpl/runner.go:179]: Having the split between both type might help troubleshooting.
```suggestion
	r.log.Debugf("Starting metadata runner with %d priority providers and %d regular providers", len(r.providers), len(r.priorityProviders))
```

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: comp/metadata/runner/runnerimpl/runner.go [hunk 6]
```diff
@@ -137,12 +176,21 @@ func (r *runnerImpl) handleProvider(p MetadataProvider) {
...
+
+	go func() {
+		for _, priorityProvider := range r.priorityProviders {
+			// Execute synchronously the priority provider
+			priorityProvider(context.Background())
```
Comment (GustavoCaso) [inline comp/metadata/runner/runnerimpl/runner.go:184]: Here is the code that ensure the HTTP request with agent V5 payload is sent first to the intake endpoint

File: comp/metadata/runner/runnerimpl/runner.go (hunk 2 above)
Comment (GustavoCaso) [reply comp/metadata/runner/runnerimpl/runner.go:181]: Should the comment on the `start` not be enough? 
```
// start is called by FX when the application starts. Lifecycle hooks are blocking and called sequencially. We should
// not block here.
```

**Review by hush-hush — APPROVED** [review]

**Pull request merged**